stores/metadata.go

```diff
@@ -820,10 +850,116 @@ func (s *SQLStore) UpdateObject(ctx context.Context, key string, o object.Object
+		if err != nil && !errors.Is(err, gorm.ErrRecordNotFound) {
+			return err
+		} else if errors.Is(err, gorm.ErrRecordNotFound) {
+			// No buffer found, create a new one.
+			return createSlabBuffer(tx, obj.ID, *partialSlab)
+		}
+
+		// We have a buffer. Sanity check it.
+		slabSize := slabSize(partialSlab.MinShards, partialSlab.TotalShards)
+		if len(buffer.Data) >= slabSize {
+			return fmt.Errorf("incomplete buffer with ID %v has no space left, this should never happen", buffer.ID)
+		}
+
+		// Create the slice.
+		remainingSpace := slabSize - len(buffer.Data)
+		slice := dbSlice{
+			DBObjectID: obj.ID,
+			DBSlabID:   buffer.DBSlabID,
+			Offset:     uint32(len(buffer.Data)),
+		}
+		if remainingSpace <= len(partialSlab.Data) {
+			slice.Length = uint32(remainingSpace)
+		} else {
+			slice.Length = uint32(len(partialSlab.Data))
+		}
+		if err := tx.Create(&slice).Error; err != nil {
+			return err
+		}
+
+		// Add the data to the buffer and remember the overflowing data.
+		var overflow []byte
+		buffer.Data = append(buffer.Data, partialSlab.Data...)
+		if len(buffer.Data) > slabSize {
+			buffer.Data, overflow = buffer.Data[:slabSize], buffer.Data[slabSize:]
+		}
+
+		// Update buffer.
+		buffer.Complete = len(buffer.Data) == slabSize
+		err = tx.Model(&dbSlabBuffer{}).
+			Where("ID", buffer.ID).
+			Updates(map[string]interface{}{
+				"complete": buffer.Complete,
+				"data":     buffer.Data,
+			}).Error
+		if err != nil {
+			return err
+		}
+
+		// If there is no overflow, we are done.
+		if len(overflow) == 0 {
+			return nil
+		}
+
+		// Otherwise, create a new buffer with a new slab and slice.
+		overflowSlab := *partialSlab
+		overflowSlab.Data = overflow
+		return createSlabBuffer(tx, obj.ID, overflowSlab)
 	})
 }
 
+func slabSize(minShards, totalShards uint8) int {
+	return int(rhpv2.SectorSize) * int(totalShards) / int(minShards)
+}
+
+func createSlabBuffer(tx *gorm.DB, objectID uint, partialSlab object.PartialSlab) error {
+	if partialSlab.TotalShards == 0 || partialSlab.MinShards == 0 {
+		return fmt.Errorf("min shards and total shards must be greater than 0: %v, %v", partialSlab.MinShards, partialSlab.TotalShards)
+	}
+	if partialSlab.MinShards > partialSlab.TotalShards {
+		return fmt.Errorf("min shards must be less than or equal to total shards: %v > %v", partialSlab.MinShards, partialSlab.TotalShards)
+	}
+	if slabSize := int(rhpv2.SectorSize) * int(partialSlab.TotalShards) / int(partialSlab.MinShards); len(partialSlab.Data) >= slabSize {
+		return fmt.Errorf("partial slab data size must be less than %v to count as a partial slab: %v", slabSize, len(partialSlab.Data))
+	}
+	key, err := object.GenerateEncryptionKey().MarshalText()
+	if err != nil {
+		return err
+	}
+	// Create a new buffer and slab.
+	return tx.Create(&dbSlabBuffer{
+		DBSlab: dbSlab{
+			Key:         key, // random placeholder key
```

Comment:
can you elaborate on the placeholder key, can it be an empty key so we know it's a placeholder?